strapi/strapi
# feat: add mapping property support for documentation plugin's open api generated docs

### What does it do?

When the documentation plugin calls cleanSchemaAttributes to convert types between attributes and openapi data types it parses dynamic zones into an array of `anyOf` components (references to openapi schemas). 
The new implementation also generates the proper discriminator property [see openapi docs](https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/) and mapping. We use the `__component` property to discriminate and the mapping grabs all compoonent names and maps them to the schemas already generated.

### Why is it needed?

Dynamic zones are reflected in the openapi docs generated as a anyOf using the list of possible components. For this to be more useful according the openapi spec we can use a discriminator to let the docs know what property is responsible for the discrimination and the mapping between the property values and schemas.

### How to test it?

Run any strap instance with a content type that has a dynamic zone. It should be generated with proper discriminator.
Taking examples from the get started example app: 

```json
"dynamic_relations": {
      "type": "array",
      "items": {
        "anyOf": [
          {
            "$ref": "#/components/schemas/BasicRelationComponent"
          },
          {
            "$ref": "#/components/schemas/BasicSimpleComponent"
          }
        ]
      },
      "discriminator": {
        "propertyName": "__component",
        "mapping": {
          "basic.relation": "#/components/schemas/BasicRelationComponent",
          "basic.simple": "#/components/schemas/BasicSimpleComponent"
        }
      }
    },
    
    
"dynamiczone": {
      "type": "array",
      "items": {
        "anyOf": [
          {
            "$ref": "#/components/schemas/BasicSimpleComponent"
          },
          {
            "$ref": "#/components/schemas/BlogTestComoComponent"
          },
          {
            "$ref": "#/components/schemas/DefaultClosingperiodComponent"
          }
        ]
      },
      "discriminator": {
        "propertyName": "__component",
        "mapping": {
          "basic.simple": "#/components/schemas/BasicSimpleComponent",
          "blog.test-como": "#/components/schemas/BlogTestComoComponent",
          "default.closingperiod": "#/components/schemas/DefaultClosingperiodComponent"
        }
      }
    },
```
